File: agent/guardian/impact_mapper.py
"""Map an ApiDiff against test coverage rows to produce impact gaps.

Coverage rows come from canary coverage (coverage-report.json) — each row
identifies which test exercises which endpoint.

Severity rules:
  CRITICAL — removed endpoint with existing tests (tests will break)
  HIGH     — added endpoint with no coverage (gap), or changed with no coverage
  MEDIUM   — changed endpoint with existing tests (silent contract drift risk)
  LOW      — added endpoint that already has coverage (rare, e.g. shared fixtures)
"""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum

from agent.guardian.diff_extractor import ApiDiff, ChangeType


class Severity(str, Enum):
    CRITICAL = "critical"
    HIGH = "high"
    MEDIUM = "medium"
    LOW = "low"

    @property
    def sort_key(self) -> int:
        return {"critical": 0, "high": 1, "medium": 2, "low": 3}[self.value]


@dataclass
class ImpactGap:
    path: str
    method: str
    operation_id: str
    change_type: ChangeType
    severity: Severity
    affected_tests: list[str] = field(default_factory=list)


def _normalize_path(path: str) -> str:
    """Normalize path parameter syntax for matching.

    Converts both `:id` and `{id}` forms to `{id}` so coverage rows
    and OpenAPI paths match regardless of which convention is used.
    """
    import re
    return re.sub(r":([a-zA-Z_][a-zA-Z0-9_]*)", r"{\1}", path)


def map_impact(diff: ApiDiff, coverage_rows: list[dict]) -> list[ImpactGap]:
    """Map diff changes to test impact gaps, sorted by severity."""
    # Build a lookup: normalized_path+method → list of test_names
    coverage: dict[tuple[str, str], list[str]] = {}
    for row in coverage_rows:
        key = (_normalize_path(row["path"]), row["method"].lower())
        coverage.setdefault(key, []).append(row["test_name"])

    gaps: list[ImpactGap] = []

    for change in diff.removed:
        key = (_normalize_path(change.path), change.method.lower())
        tests = coverage.get(key, [])
        gaps.append(ImpactGap(
            path=change.path,
            method=change.method,
            operation_id=change.operation_id,
            change_type=ChangeType.REMOVED,
            severity=Severity.CRITICAL if tests else Severity.HIGH,
            affected_tests=tests,
        ))

    for change in diff.added:
        key = (_normalize_path(change.path), change.method.lower())
        tests = coverage.get(key, [])
        gaps.append(ImpactGap(
            path=change.path,
            method=change.method,
            operation_id=change.operation_id,
            change_type=ChangeType.ADDED,
            severity=Severity.LOW if tests else Severity.HIGH,
            affected_tests=tests,
        ))

    for change in diff.changed:
        key = (_normalize_path(change.path), change.method.lower())
        tests = coverage.get(key, [])
        gaps.append(ImpactGap(
            path=change.path,
            method=change.method,
            operation_id=change.operation_id,
            change_type=ChangeType.CHANGED,
            severity=Severity.MEDIUM if tests else Severity.HIGH,
            affected_tests=tests,
        ))

    return sorted(gaps, key=lambda g: g.severity.sort_key)

File: ts/src/guardian/impact-mapper.ts
/**
 * Map an `ApiDiff` against test coverage rows to produce impact gaps.
 *
 * Faithful TypeScript port of `agent/guardian/impact_mapper.py`. Coverage rows
 * come from canary coverage (coverage-report.json) — each row identifies which
 * test exercises which endpoint.
 *
 * Severity rules:
 *   CRITICAL — removed endpoint with existing tests (tests will break)
 *   HIGH     — added endpoint with no coverage (gap), or changed with no coverage
 *   MEDIUM   — changed endpoint with existing tests (silent contract drift risk)
 *   LOW      — added endpoint that already has coverage (rare, shared fixtures)
 */

import { ApiDiff, ChangeType } from './diff-extractor.js';

/** Python: `Severity(str, Enum)`. */
export enum Severity {
  CRITICAL = 'critical',
  HIGH = 'high',
  MEDIUM = 'medium',
  LOW = 'low',
}

/**
 * Python: `Severity.sort_key` (ascending — CRITICAL sorts first). The canonical
 * severity ordering; reached from outside this module through
 * {@link severitySortKey} rather than directly (#544).
 */
const SEVERITY_SORT_KEY: Record<Severity, number> = {
  [Severity.CRITICAL]: 0,
  [Severity.HIGH]: 1,
  [Severity.MEDIUM]: 2,
  [Severity.LOW]: 3,
};

/** Python: `Severity.sort_key` accessor. */
export function severitySortKey(severity: Severity): number {
  return SEVERITY_SORT_KEY[severity];
}

/** A coverage row from coverage-report.json. */
export interface CoverageRow {
  path: string;
  method: string;
  test_name: string;
  [key: string]: unknown;
}

/** Python: `ImpactGap` dataclass. */
export class ImpactGap {
  path: string;
  method: string;
  // `string | null`: carries through an EndpointChange.operation_id that was a
  // present-null in the source spec (Python assigns change.operation_id as-is).
  operation_id: string | null;
  change_type: ChangeType;
  severity: Severity;
  affected_tests: string[];

  constructor(init: {
    path: string;
    method: string;
    operation_id: string | null;
    change_type: ChangeType;
    severity: Severity;
    affected_tests?: string[];
  }) {
    this.path = init.path;
    this.method = init.method;
    this.operation_id = init.operation_id;
    this.change_type = init.change_type;
    this.severity = init.severity;
    this.affected_tests = init.affected_tests ?? [];
  }
}

/**
 * Python: `_normalize_path`. Normalize path parameter syntax for matching:
 * converts both `:id` and `{id}` forms to `{id}` so coverage rows and OpenAPI
 * paths match regardless of which convention is used.
 */
function normalizePath(path: string): string {
  return path.replace(/:([a-zA-Z_][a-zA-Z0-9_]*)/g, '{$1}');
}

function coverageKey(path: string, method: string): string {
  return JSON.stringify([normalizePath(path), method.toLowerCase()]);
}

/**
 * Python: `map_impact`. Map diff changes to test impact gaps, sorted by
 * severity (stable — CRITICAL first).
 */
export function mapImpact(
  diff: ApiDiff,
  coverageRows: CoverageRow[],
): ImpactGap[] {
  // Build a lookup: normalized_path+method -> list of test_names.
  const coverage = new Map<string, string[]>();
  for (const row of coverageRows) {
    const key = coverageKey(row.path, row.method);
    const existing = coverage.get(key);
    if (existing) {
      existing.push(row.test_name);
    } else {
      coverage.set(key, [row.test_name]);
    }
  }

  const gaps: ImpactGap[] = [];

  for (const change of diff.removed) {
    const tests = coverage.get(coverageKey(change.path, change.method)) ?? [];
    gaps.push(
      new ImpactGap({
        path: change.path,
        method: change.method,
        operation_id: change.operation_id,
        change_type: ChangeType.REMOVED,
        severity: tests.length ? Severity.CRITICAL : Severity.HIGH,
        affected_tests: tests,
      }),
    );
  }

  for (const change of diff.added) {
    const tests = coverage.get(coverageKey(change.path, change.method)) ?? [];
    gaps.push(
      new ImpactGap({
        path: change.path,
        method: change.method,
        operation_id: change.operation_id,
        change_type: ChangeType.ADDED,
        severity: tests.length ? Severity.LOW : Severity.HIGH,
        affected_tests: tests,
      }),
    );
  }

  for (const change of diff.changed) {
    const tests = coverage.get(coverageKey(change.path, change.method)) ?? [];
    gaps.push(
      new ImpactGap({
        path: change.path,
        method: change.method,
        operation_id: change.operation_id,
        change_type: ChangeType.CHANGED,
        severity: tests.length ? Severity.MEDIUM : Severity.HIGH,
        affected_tests: tests,
      }),
    );
  }

  // Python `sorted(...)` returns a new list and is stable.
  return [...gaps].sort(
    (a, b) => severitySortKey(a.severity) - severitySortKey(b.severity),
  );
}
